File: scripts/merge-benchmark-summaries.py
#!/usr/bin/env python3

import argparse
import json
import pathlib
from datetime import datetime, UTC


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--output", required=True)
    parser.add_argument("summary_files", nargs="+")
    return parser.parse_args()


def median(values: list[float]) -> float:
    if not values:
        return 0
    sorted_values = sorted(values)
    midpoint = len(sorted_values) // 2
    if len(sorted_values) % 2 == 1:
        return sorted_values[midpoint]
    return (sorted_values[midpoint - 1] + sorted_values[midpoint]) / 2


def rate(num: int, den: int) -> float:
    if den == 0:
        return 0
    return num / den


def count_preferred(primary: list[str], secondary: list[str], fallback: int = 0) -> int:
    if primary:
        return len(primary)
    if secondary:
        return len(secondary)
    return fallback


def normalize_string_list(value: object) -> list[str]:
    if not isinstance(value, list):
        return []
    normalized: list[str] = []
    for item in value:
        if isinstance(item, str):
            cleaned = item.strip()
            if cleaned:
                normalized.append(cleaned)
    return normalized


def merge_string_lists(summary_payloads: list[dict], field_name: str) -> list[str]:
    merged: list[str] = []
    seen: set[str] = set()
    for payload in summary_payloads:
        for value in normalize_string_list(payload.get(field_name)):
            if value in seen:
                continue
            seen.add(value)
            merged.append(value)
    return merged


def task_ids(tasks: list[dict]) -> list[str]:
    return [str(task["task_id"]) for task in tasks if isinstance(task.get("task_id"), str)]


def task_paths(tasks: list[dict]) -> list[str]:
    paths: list[str] = []
    for task in tasks:
        path = task.get("task_path") or task.get("task_file") or task.get("path")
        if isinstance(path, str) and path.strip():
            paths.append(path.strip())
    return paths


def merge_unique(primary: list[str], secondary: list[str]) -> list[str]:
    merged: list[str] = []
    seen: set[str] = set()
    for value in [*primary, *secondary]:
        if value in seen:
            continue
        seen.add(value)
        merged.append(value)
    return merged


def merge_summaries(summary_payloads: list[dict]) -> dict:
    if not summary_payloads:
        raise ValueError("at least one summary is required")

    first = summary_payloads[0]
    tasks = []
    configured_tasks = 0
    executed_tasks = 0
    for payload in summary_payloads:
        configured_tasks += int(payload["totals"]["configured_tasks"])
        executed_tasks += int(payload["totals"]["executed_tasks"])
        tasks.extend(payload.get("tasks", []))

    selected_task_ids = merge_string_lists(summary_payloads, "selected_task_ids")
    selected_task_paths = merge_string_lists(summary_payloads, "selected_task_paths")
    executed_task_ids = merge_string_lists(summary_payloads, "executed_task_ids")
    executed_task_paths = merge_string_lists(summary_payloads, "executed_task_paths")
    unexecuted_task_ids = merge_string_lists(summary_payloads, "unexecuted_task_ids")
    unexecuted_task_paths = merge_string_lists(summary_payloads, "unexecuted_task_paths")
    unresolved_task_ids = merge_string_lists(summary_payloads, "unresolved_task_ids")
    unresolved_task_paths = merge_string_lists(summary_payloads, "unresolved_task_paths")

    total = len(tasks)
    failed_tasks = [task for task in tasks if task.get("status") != "passed"]
    passed = len([task for task in tasks if task["status"] == "passed"])
    clean_passed = len(
        [
            task
            for task in tasks
            if task["status"] == "passed"
            and task.get("recovered_nonzero_exit") is not True
            and (task.get("summary_repaired_by") or "none") == "none"
        ]
    )
    completed = len([task for task in tasks if task.get("completed") is True])
    verification_required = len([task for task in tasks if task.get("verification_required") is True])
    tests_run = len([task for task in tasks if task.get("tests_run") is True])
    tests_passed = len([task for task in tasks if task.get("tests_passed") is True])
    review_required = len([task for task in tasks if task.get("review_required") is True])
    review_present = len([task for task in tasks if task.get("review_present") is True])
    docs_required = len([task for task in tasks if task.get("docs_required") is True])
    docs_updated = len([task for task in tasks if task.get("docs_updated") is True])
    recovered_tasks = len([task for task in tasks if task.get("recovered_nonzero_exit") is True])
    timeout_recovered = len([task for task in tasks if task.get("timeout_recovered") is True])
    max_turns_recovered = len([task for task in tasks if task.get("max_turns_recovered") is True])
    summary_repaired = len([task for task in tasks if (task.get("summary_repaired_by") or "none") != "none"])
    policy_violations = sum(int(task.get("policy_violations", 0)) for task in tasks)
    tool_failures = sum(int(task.get("tool_failures", 0)) for task in tasks)
    selected_total = count_preferred(selected_task_ids, selected_task_paths, configured_tasks)
    unexecuted_total = count_preferred(
        unexecuted_task_ids,
        unexecuted_task_paths,
        max(selected_total - executed_tasks, 0),
    )
    unresolved_total = (
        count_preferred(
            unresolved_task_ids,
            unresolved_task_paths,
            len([task for task in tasks if task.get("status") != "passed"]) + unexecuted_total,
        )
    )

    if not executed_task_ids:
        executed_task_ids = task_ids(tasks)
    if not executed_task_paths:
        executed_task_paths = task_paths(tasks)
    if not unresolved_task_ids:
        unresolved_task_ids = merge_unique(task_ids(failed_tasks), unexecuted_task_ids)
    if not unresolved_task_paths:
        unresolved_task_paths = merge_unique(task_paths(failed_tasks), unexecuted_task_paths)

    return {
        "schema_version": first["schema_version"],
        "mode": first["mode"],
        "runner": first["runner"],
        "generated_at": datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "source_ref": first["source_ref"],
        "source_sha": first["source_sha"],
        "task_glob": first["task_glob"],
        "selected_task_ids": selected_task_ids,
        "selected_task_paths": selected_task_paths,
        "executed_task_ids": executed_task_ids,
        "executed_task_paths": executed_task_paths,
        "unexecuted_task_ids": unexecuted_task_ids,
        "unexecuted_task_paths": unexecuted_task_paths,
        "unresolved_task_ids": unresolved_task_ids,
        "unresolved_task_paths": unresolved_task_paths,
        "totals": {
            "configured_tasks": configured_tasks,
            "selected_tasks": selected_total,
            "executed_tasks": executed_tasks,
            "unexecuted_tasks": unexecuted_total,
            "unresolved_tasks": unresolved_total,
            "tasks": executed_tasks,
            "passed": passed,
            "clean_passed": clean_passed,
            "completed": completed,
            "verification_required": verification_required,
            "tests_run": tests_run,
            "tests_passed": tests_passed,
            "review_required": review_required,
            "review_present": review_present,
            "docs_required": docs_required,
            "docs_updated": docs_updated,
            "recovered_tasks": recovered_tasks,
            "timeout_recovered": timeout_recovered,
            "max_turns_recovered": max_turns_recovered,
            "summary_repaired": summary_repaired,
            "policy_violations": policy_violations,
            "tool_failures": tool_failures,
        },
        "rates": {
            "task_pass_rate": rate(passed, total),
            "clean_pass_rate": rate(clean_passed, total),
            "completion_rate": rate(completed, total),
            "verification_rate": rate(len([task for task in tasks if (task.get("verification_required") is False) or (task.get("tests_run") is True)]), total),
            "verification_pass_rate": rate(len([task for task in tasks if (task.get("verification_required") is False) or (task.get("tests_passed") is True)]), total),
            "review_compliance_rate": rate(len([task for task in tasks if (task.get("review_required") is False) or (task.get("review_present") is True)]), total),
            "docs_compliance_rate": rate(len([task for task in tasks if (task.get("docs_required") is False) or (task.get("docs_updated") is True)]), total),
            "recovered_task_rate": rate(recovered_tasks, total),
            "summary_repair_rate": rate(summary_repaired, total),
            "execution_coverage_rate": rate(executed_tasks, configured_tasks),
            "unexecuted_rate": rate(unexecuted_total, selected_total),
            "unresolved_rate": rate(unresolved_total, selected_total),
        },
        "median_runtime_seconds": median([float(task["runtime_seconds"]) for task in tasks]),
        "tasks": tasks,
    }


def main() -> None:
    args = parse_args()
    payloads = [
        json.loads(pathlib.Path(summary_file).read_text(encoding="utf-8"))
        for summary_file in args.summary_files
    ]
    merged = merge_summaries(payloads)
    pathlib.Path(args.output).write_text(json.dumps(merged, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


if __name__ == "__main__":
    main()

File: scripts/merge-benchmark-summaries.mjs
#!/usr/bin/env node
// merge-benchmark-summaries: merge shard summaries into one aggregate summary.
// Node port of scripts/merge-benchmark-summaries.py — no Python.
import { isMain, readJson, writeJson, median, rate, normalizeStringList } from './bench/lib.mjs';

export { median, rate, normalizeStringList };

export function countPreferred(primary, secondary, fallback = 0) {
  if (primary && primary.length) return primary.length;
  if (secondary && secondary.length) return secondary.length;
  return fallback;
}

export function mergeStringLists(summaryPayloads, fieldName) {
  const merged = [];
  const seen = new Set();
  for (const payload of summaryPayloads) {
    for (const v of normalizeStringList(payload[fieldName])) {
      if (!seen.has(v)) { seen.add(v); merged.push(v); }
    }
  }
  return merged;
}

export function taskIds(tasks) {
  return tasks.map((t) => t.task_id).filter((id) => typeof id === 'string');
}

export function taskPaths(tasks) {
  const paths = [];
  for (const t of tasks) {
    const p = t.task_path || t.task_file || t.path;
    if (typeof p === 'string' && p.trim()) paths.push(p.trim());
  }
  return paths;
}

export function mergeUnique(primary, secondary) {
  const merged = [];
  const seen = new Set();
  for (const v of [...primary, ...secondary]) {
    if (!seen.has(v)) { seen.add(v); merged.push(v); }
  }
  return merged;
}

function countWhere(tasks, pred) {
  return tasks.filter(pred).length;
}

export function mergeSummaries(summaryPayloads) {
  if (!summaryPayloads || summaryPayloads.length === 0) {
    throw new Error('at least one summary is required');
  }
  const first = summaryPayloads[0];
  const tasks = [];
  let configuredTasks = 0;
  let executedTasks = 0;
  for (const p of summaryPayloads) {
    configuredTasks += Number(p.totals.configured_tasks);
    executedTasks += Number(p.totals.executed_tasks);
    for (const t of (p.tasks || [])) tasks.push(t);
  }

  const selectedTaskIds = mergeStringLists(summaryPayloads, 'selected_task_ids');
  const selectedTaskPaths = mergeStringLists(summaryPayloads, 'selected_task_paths');
  let executedTaskIds = mergeStringLists(summaryPayloads, 'executed_task_ids');
  let executedTaskPaths = mergeStringLists(summaryPayloads, 'executed_task_paths');
  const unexecutedTaskIds = mergeStringLists(summaryPayloads, 'unexecuted_task_ids');
  const unexecutedTaskPaths = mergeStringLists(summaryPayloads, 'unexecuted_task_paths');
  let unresolvedTaskIds = mergeStringLists(summaryPayloads, 'unresolved_task_ids');
  let unresolvedTaskPaths = mergeStringLists(summaryPayloads, 'unresolved_task_paths');

  const total = tasks.length;
  const failedTasks = tasks.filter((t) => t.status !== 'passed');
  const passed = countWhere(tasks, (t) => t.status === 'passed');
  const cleanPassed = countWhere(tasks, (t) => t.status === 'passed' && t.recovered_nonzero_exit !== true && (t.summary_repaired_by || 'none') === 'none');
  const completed = countWhere(tasks, (t) => t.completed === true);
  const verificationRequired = countWhere(tasks, (t) => t.verification_required === true);
  const testsRun = countWhere(tasks, (t) => t.tests_run === true);
  const testsPassed = countWhere(tasks, (t) => t.tests_passed === true);
  const reviewRequired = countWhere(tasks, (t) => t.review_required === true);
  const reviewPresent = countWhere(tasks, (t) => t.review_present === true);
  const docsRequired = countWhere(tasks, (t) => t.docs_required === true);
  const docsUpdated = countWhere(tasks, (t) => t.docs_updated === true);
  const recoveredTasks = countWhere(tasks, (t) => t.recovered_nonzero_exit === true);
  const timeoutRecovered = countWhere(tasks, (t) => t.timeout_recovered === true);
  const maxTurnsRecovered = countWhere(tasks, (t) => t.max_turns_recovered === true);
  const summaryRepaired = countWhere(tasks, (t) => (t.summary_repaired_by || 'none') !== 'none');
  const policyViolations = tasks.reduce((s, t) => s + Number(t.policy_violations || 0), 0);
  const toolFailures = tasks.reduce((s, t) => s + Number(t.tool_failures || 0), 0);

  const selectedTotal = countPreferred(selectedTaskIds, selectedTaskPaths, configuredTasks);
  const unexecutedTotal = countPreferred(unexecutedTaskIds, unexecutedTaskPaths, Math.max(selectedTotal - executedTasks, 0));
  const unresolvedTotal = countPreferred(unresolvedTaskIds, unresolvedTaskPaths, failedTasks.length + unexecutedTotal);

  if (!executedTaskIds.length) executedTaskIds = taskIds(tasks);
  if (!executedTaskPaths.length) executedTaskPaths = taskPaths(tasks);
  if (!unresolvedTaskIds.length) unresolvedTaskIds = mergeUnique(taskIds(failedTasks), unexecutedTaskIds);
  if (!unresolvedTaskPaths.length) unresolvedTaskPaths = mergeUnique(taskPaths(failedTasks), unexecutedTaskPaths);

  return {
    schema_version: first.schema_version,
    mode: first.mode,
    runner: first.runner,
    generated_at: new Date().toISOString().replace(/\.\d{3}Z$/, 'Z'),
    source_ref: first.source_ref,
    source_sha: first.source_sha,
    task_glob: first.task_glob,
    selected_task_ids: selectedTaskIds,
    selected_task_paths: selectedTaskPaths,
    executed_task_ids: executedTaskIds,
    executed_task_paths: executedTaskPaths,
    unexecuted_task_ids: unexecutedTaskIds,
    unexecuted_task_paths: unexecutedTaskPaths,
    unresolved_task_ids: unresolvedTaskIds,
    unresolved_task_paths: unresolvedTaskPaths,
    totals: {
      configured_tasks: configuredTasks,
      selected_tasks: selectedTotal,
      executed_tasks: executedTasks,
      unexecuted_tasks: unexecutedTotal,
      unresolved_tasks: unresolvedTotal,
      tasks: executedTasks,
      passed,
      clean_passed: cleanPassed,
      completed,
      verification_required: verificationRequired,
      tests_run: testsRun,
      tests_passed: testsPassed,
      review_required: reviewRequired,
      review_present: reviewPresent,
      docs_required: docsRequired,
      docs_updated: docsUpdated,
      recovered_tasks: recoveredTasks,
      timeout_recovered: timeoutRecovered,
      max_turns_recovered: maxTurnsRecovered,
      summary_repaired: summaryRepaired,
      policy_violations: policyViolations,
      tool_failures: toolFailures,
    },
    rates: {
      task_pass_rate: rate(passed, total),
      clean_pass_rate: rate(cleanPassed, total),
      completion_rate: rate(completed, total),
      verification_rate: rate(countWhere(tasks, (t) => t.verification_required === false || t.tests_run === true), total),
      verification_pass_rate: rate(countWhere(tasks, (t) => t.verification_required === false || t.tests_passed === true), total),
      review_compliance_rate: rate(countWhere(tasks, (t) => t.review_required === false || t.review_present === true), total),
      docs_compliance_rate: rate(countWhere(tasks, (t) => t.docs_required === false || t.docs_updated === true), total),
      recovered_task_rate: rate(recoveredTasks, total),
      summary_repair_rate: rate(summaryRepaired, total),
      execution_coverage_rate: rate(executedTasks, configuredTasks),
      unexecuted_rate: rate(unexecutedTotal, selectedTotal),
      unresolved_rate: rate(unresolvedTotal, selectedTotal),
    },
    median_runtime_seconds: median(tasks.map((t) => Number(t.runtime_seconds))),
    tasks,
  };
}

function parseArgs(argv) {
  const out = { output: null, summaryFiles: [] };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--output') { out.output = argv[++i]; }
    else { out.summaryFiles.push(a); }
  }
  if (!out.output) { process.stderr.write('Usage: merge-benchmark-summaries.mjs --output FILE SUMMARY... \n'); process.exit(2); }
  if (out.summaryFiles.length === 0) { process.stderr.write('at least one summary is required\n'); process.exit(2); }
  return out;
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const payloads = args.summaryFiles.map((f) => readJson(f));
  const merged = mergeSummaries(payloads);
  writeJson(args.output, merged);
}

if (isMain(import.meta.url)) {
  main();
}